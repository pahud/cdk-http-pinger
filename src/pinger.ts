import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import * as logs from '@aws-cdk/aws-logs';
import { Construct, CustomResource } from '@aws-cdk/core';
import * as cr from '@aws-cdk/custom-resources';

export interface PingerProps {
  readonly parameter?: { [key: string]: string };
  readonly url: string;
  /**
   * optional entry file
   */
  readonly entry?: string;
}

export class Pinger extends Construct {
  readonly resource: CustomResource;
  readonly httpStatus: string;
  readonly htmlTitle: string;
  readonly url: string;
  readonly body: string;
  constructor(scope: Construct, id: string, props: PingerProps) {
    super(scope, id);

    this.url = props.url;
    const onEvent = new NodejsFunction(this, 'ProviderFunc', {
      entry: props.entry ?? path.join(__dirname, './lambda/index.js'),
      handler: 'onEvent',
      runtime: lambda.Runtime.NODEJS_14_X,
      bundling: {
        define: {
          'process.env.PARAMETER': jsonStringifiedBundlingDefinition(props.parameter ?? {}),
        },
      },
    });

    const myProvider = new cr.Provider(this, 'Provider', {
      onEventHandler: onEvent,
      logRetention: logs.RetentionDays.ONE_DAY,
    });

    const pingerResource = new CustomResource(this, 'PingerResource', {
      serviceToken: myProvider.serviceToken,
      resourceType: 'Custom::Pinger',
      properties: {
        url: props.url,
      },
    });
    this.resource = pingerResource;
    this.httpStatus = pingerResource.getAtt('status').toString();
    this.htmlTitle = pingerResource.getAtt('title').toString();
    this.body = pingerResource.getAtt('body').toString();
  }
}

function jsonStringifiedBundlingDefinition(value: any): string {
  return JSON.stringify(value)
    .replace(/"/g, '\\"')
    .replace(/,/g, '\\,');
}
