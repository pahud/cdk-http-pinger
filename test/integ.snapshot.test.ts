import '@aws-cdk/assert/jest';
import * as path from 'path';
import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { Pinger } from '../src/pinger';

test('integ snapshot validation', () => {
  const app = new cdk.App();

  const stack = new cdk.Stack(app, 'my-stack-dev');

  const pinger = new Pinger(stack, 'Pinger', {
    url: 'https://aws.amazon.com',
    /**
     * As jest will rm -rf ./lib, we need this to redirect the entry to src/index.ts
     */
    entry: path.join(__dirname, '../src/lambda/index.ts'),
  });

  new cdk.CfnOutput(stack, 'HttpStatus', { value: pinger.httpStatus });
  new cdk.CfnOutput(stack, 'HtmlTitle', { value: pinger.htmlTitle });
  new cdk.CfnOutput(stack, 'URL', { value: pinger.url });
  new cdk.CfnOutput(stack, 'Body', { value: pinger.body });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();

  app.synth();

});
