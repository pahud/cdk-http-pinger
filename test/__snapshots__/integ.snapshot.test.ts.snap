// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integ snapshot validation 1`] = `
Object {
  "Outputs": Object {
    "Body": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "PingerPingerResourceDE344ADF",
          "body",
        ],
      },
    },
    "HtmlTitle": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "PingerPingerResourceDE344ADF",
          "title",
        ],
      },
    },
    "HttpStatus": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "PingerPingerResourceDE344ADF",
          "status",
        ],
      },
    },
    "URL": Object {
      "Value": "https://aws.amazon.com",
    },
  },
  "Parameters": Object {
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24ArtifactHashBA91B77F": Object {
      "Description": "Artifact hash for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5": Object {
      "Description": "S3 bucket for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861": Object {
      "Description": "S3 key for asset version \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters8f243f7445fa2b77a361686d50ee88a38df1ce4e640cada90bb3bb214cf4f6a2ArtifactHashE066F4CF": Object {
      "Description": "Artifact hash for asset \\"8f243f7445fa2b77a361686d50ee88a38df1ce4e640cada90bb3bb214cf4f6a2\\"",
      "Type": "String",
    },
    "AssetParameters8f243f7445fa2b77a361686d50ee88a38df1ce4e640cada90bb3bb214cf4f6a2S3BucketC49B8495": Object {
      "Description": "S3 bucket for asset \\"8f243f7445fa2b77a361686d50ee88a38df1ce4e640cada90bb3bb214cf4f6a2\\"",
      "Type": "String",
    },
    "AssetParameters8f243f7445fa2b77a361686d50ee88a38df1ce4e640cada90bb3bb214cf4f6a2S3VersionKey64A0093F": Object {
      "Description": "S3 key for asset version \\"8f243f7445fa2b77a361686d50ee88a38df1ce4e640cada90bb3bb214cf4f6a2\\"",
      "Type": "String",
    },
    "AssetParametersc691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49cArtifactHash627DAAA7": Object {
      "Description": "Artifact hash for asset \\"c691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49c\\"",
      "Type": "String",
    },
    "AssetParametersc691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49cS3BucketEAC9DD43": Object {
      "Description": "S3 bucket for asset \\"c691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49c\\"",
      "Type": "String",
    },
    "AssetParametersc691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49cS3VersionKeyDD9AE9E7": Object {
      "Description": "S3 key for asset version \\"c691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49c\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PingerPingerResourceDE344ADF": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "PingerProviderframeworkonEventB03B6EDA",
            "Arn",
          ],
        },
        "url": "https://aws.amazon.com",
      },
      "Type": "Custom::Pinger",
      "UpdateReplacePolicy": "Delete",
    },
    "PingerProviderFunc00EB0D69": Object {
      "DependsOn": Array [
        "PingerProviderFuncServiceRole61DD47CB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters8f243f7445fa2b77a361686d50ee88a38df1ce4e640cada90bb3bb214cf4f6a2S3BucketC49B8495",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8f243f7445fa2b77a361686d50ee88a38df1ce4e640cada90bb3bb214cf4f6a2S3VersionKey64A0093F",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8f243f7445fa2b77a361686d50ee88a38df1ce4e640cada90bb3bb214cf4f6a2S3VersionKey64A0093F",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "PingerProviderFuncServiceRole61DD47CB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "PingerProviderFuncServiceRole61DD47CB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PingerProviderframeworkonEventB03B6EDA": Object {
      "DependsOn": Array [
        "PingerProviderframeworkonEventServiceRoleDefaultPolicy13F4346C",
        "PingerProviderframeworkonEventServiceRoleA8D3E90C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersc691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49cS3BucketEAC9DD43",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49cS3VersionKeyDD9AE9E7",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49cS3VersionKeyDD9AE9E7",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "AWS CDK resource provider framework - onEvent (my-stack-dev/Pinger/Provider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "PingerProviderFunc00EB0D69",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "PingerProviderframeworkonEventServiceRoleA8D3E90C",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "PingerProviderframeworkonEventLogRetentionF394D8B3": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "PingerProviderframeworkonEventB03B6EDA",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "PingerProviderframeworkonEventServiceRoleA8D3E90C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PingerProviderframeworkonEventServiceRoleDefaultPolicy13F4346C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PingerProviderFunc00EB0D69",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PingerProviderframeworkonEventServiceRoleDefaultPolicy13F4346C",
        "Roles": Array [
          Object {
            "Ref": "PingerProviderframeworkonEventServiceRoleA8D3E90C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
